### PR TITLE
MAINT: prepare for SciPy 1.11.2

### PR DESCRIPTION
**Pull request by tylerjereddy**
SciPy `1.11.2` would tentatively include Python `3.12` and `musllinux` wheels, and a bunch of bug fixes.

Possible TODOs:

- [x] wait for the next NumPy release, which should have Python `3.12` wheels (I think this makes sense to do, and I believe upstream is working hard on this) -- if we'd prefer to delay `3.12` wheels to `1.11.3` let me know below
- [x] the CI/wheel infrastructure was updated/backported, so decent chance we may see some surprises with `[wheel build]` -- I may decide to do that here once rather than a bit later to reduce likelihood of a surprise during the release process proper (the net resource cost could be a fair bit higher if only discovered during rel process and then I need to patch/start over)
- [x] should we try to wrap up gh-18781? my impression from taking a stab at this in gh-19034 is that it is a can of worms because of potential `f2py` involvement...
- [x] we still have some open PRs with backport label--may be able to get some of those merged while we wait a little on NumPy `3.12` wheels and so on
- [ ] I feel like I may be missing a Cython stdout build bug of some sort that was affecting `networkx` or `scikit-image`? Maybe it is dealt with by something here, but that may be good to confirm, or maybe I'm just misremembering.

Backports Included so far:

- gh-18644
- gh-18785
- gh-18790
- gh-18804
- gh-18810
- gh-18825
- gh-18831
- gh-18859
- gh-18896
- gh-18937
- gh-18985
- gh-18999
- gh-19025
- gh-19027 
- gh-17918
- gh-18809
- gh-19004
- gh-19034
- gh-19054

### CODE REVIEW

**Comment by tylerjereddy** [comment]
Some fun failures in CI, adding WIP to title for now.

**Comment by rgommers** [comment]
> wait for the next NumPy release, which should have Python `3.12` wheels (I think this makes sense to do, and I believe upstream is working hard on this) -- if we'd prefer to delay `3.12` wheels to `1.11.3` let me know below

I think it's best to wait here, shouldn't take much longer.

File: .github/workflows/wheels.yml
Comment (rgommers) [inline .github/workflows/wheels.yml:126]: I think it'd be better to move to `2.15` here, that's the first release with proper Python 3.12 support without use of `CIBW_PRERELEASE_PYTHONS`. @andyfaff WDYT?

File: ci/cirrus_wheels.yml
Comment (rgommers) [inline ci/cirrus_wheels.yml:3]: 2.15.0 here too?

**Review by rgommers — APPROVED** [review]
All backports LGTM. Most of the open questions are nice to take along, but can be deferred it looks like. I don't have a memory of the Cython bug you mention.

File: ci/cirrus_wheels.yml
Comment (andyfaff) [reply ci/cirrus_wheels.yml:3]: cibuildwheel makes frequent releases, and they're well tested. I think bumping versions to the latest holds little risks. You're seeing 2.14.1 because that's the latest version that was around when I made the original commit.

**Comment by dschmitz89** [comment]
If time permits, could this recent fix https://github.com/scipy/scipy/pull/19004 also be included?

**Comment by tylerjereddy** [comment]
> If time permits, could this recent fix https://github.com/scipy/scipy/pull/19004 also be included?

Yup, already on my list. I imagine we may have a few more shims before this gets merged.

**Comment by h-vetinari** [comment]
Unless and until we get the DeprecationWarnings with numpy 1.25 down to zero on the 1.11 branch, I think we should lower the maximum numpy version here from 1.28 to 1.27, see [this comment](https://github.com/scipy/scipy/pull/19034#issuecomment-1674093550).

**Comment by tylerjereddy** [comment]
I think we're going to hit `0` warnings, there's just the 4-line patch in gh-19034 left to review.

**Comment by tylerjereddy** [comment]
Surveying some of the CI failures (if only to keep myself organized):

- GitHub Actions `musllinux`: `tools/wheels/cibw_before_build_linux.sh: line 8: can't create --nightly/LICENSE.txt: nonexistent directory`
  - this may require me to cherry-pick a subset of gh-18889, where nightly-related variable handling shims were applied to `tools/wheels/cibw_before_build_linux.sh` fairly recently
- GHA build with gcc-8 on Linux:  `ERROR Missing dependencies:
	Cython<3.0,>=0.29.35
	pybind11<2.11.0,>=2.10.4
	wheel<0.41.0
`
  - we can probably just pin the Cython version in `.github/workflows/linux_meson.yml` for this job and move on; it disables build isolation on a release branch
- (Intentional) `setup.py` build/wheel install on windows with Python `3.10`-- `error LNK2001: unresolved external symbol wrap_PyUFunc_getfperr` and maybe warnings could be related `WARN: could not find library 'sc_cephes' in directories`
  - not sure here just yet, but IIRC Cython `3.0.0` with `setup.py` is asking for trouble, and perhaps we can just pin that in this job and still provide what conda-forge still needs for residual Windows + `setup.py` usage?
-  Linux `meson` "pre-release": `No matching distribution found for numpy<1.28.0`
    -  I think a `pip --pre` type pre-release install is not finding a NumPy nightly wheel with a version that isn't `2.0.0-something`--I'm inclined to just let that one keep failing--trying to support the `2.0.0-dev` series on the maintenance branch sounds like a recipe for unsustainable trouble
-  Linux 32-bit: need to backport gh-18809

**Comment by tylerjereddy** [comment]
I tried to apply those smaller patches along with another batch of backports detailed in the updated original comment above. I can already see that the `gcc-8` job will require a few more version pins I think. Skipping cirrus for now, those jobs were already passing.

**Comment by tylerjereddy** [comment]
Maybe also remind me--do we want to wait for a *stable* release of NumPy that has wheels for `3.12` or are we content with the pre-release `1.26.0b1` already out? Ideally stable I suppose, right?

**Comment by tylerjereddy** [comment]
There is only one CI failure left, and it is an unrelated `datasets` internet connectivity issue, so that's progress for the regular CI. The next thing I'll try is a test wheel build on this branch--I want to confirm no surprises, and I'm perhaps slightly skeptical that the `3.12` wheels will "just build" without a tweak or two (i.e., pre-release req for NumPy somewhere perhaps, etc.).

**Comment by tylerjereddy** [comment]
Yeah, I can already see a wheel failure trying to use `numpy-2.0.0.dev0`, which is wrong, so some shims needed still, though could be quite minor.

**Comment by rgommers** [comment]
In case it helps and you didn't see it: https://github.com/numpy/numpy/issues/23808#issuecomment-1678677094 and the comment a bit higher about `PIP_PRE` may be useful.

**Comment by tylerjereddy** [comment]
Summary of observed issues with Python `3.12` wheels so far:

1) `musllinux`
  - 1 test failure: `TestFractionalMatrixPower.test_al_mohy_higham_2012_experiment_1` at `scipy/linalg/tests/test_matfuncs.py:559` may need a tolerance bump for relative max diff of `7.29361661e-12`
  - 44 warnings of the form: `python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=6332) is multi-threaded, use of fork() may lead to deadlocks in the child.`
2) Cirrus Linux ARM: `datasets` connectivity issues were observed with another Python version; also a brownout warning re: exceeding our free limit stopped some jobs--I'll try restarting, I need to see if there's a `3.12` issue here.

The other jobs, including the other `3.12` wheels, appear to have worked "ok" (I didn't inspect them for similar warnings, but I'm assuming we can live with those in our first `3.12` iteration).

**Comment by tylerjereddy** [comment]
@rgommers I thought of one more question--do we want to tackle: https://numpy.org/doc/stable/release/1.25.0-notes.html#compiling-against-the-numpy-c-api-is-now-backwards-compatible-by-default

That might be one thing too many to do here (a `pyproject.toml` readjustment), increasing the likelihood of a mistake by a bit too much perhaps, but on the other hand we do intend to take advantage of that at some point?

**Comment by tylerjereddy** [comment]
@IAlibay pointed out that when switching to that `1.25+` approach for MDA, there were some snags with `conda-forge` cc @h-vetinari perhaps.

**Comment by h-vetinari** [comment]
> @IAlibay pointed out that when switching to that `1.25+` approach for MDA, there were some snags with `conda-forge`

If there are issues with conda-forge packages, could you please open an issue on the respective feedstock and ping me there?
Note that conda-forge still builds for the lowest (NEP29-compatible) numpy by default, we haven't switched to the 1.25+ approach yet (and we haven't discussed, much less figured out, how we'll do that).

**Comment by IAlibay** [comment]
Thanks @h-vetinari - To clarify the issue that @tylerjereddy mentions isn't with the scipy feedstock but pretty much the following:

>  much less figured out, how we'll do that

 We're in the middle of trying to do the jump from using oldest compatible and working out that it appears to be a much harder task than anticipated (at least my naive understanding of the conda-forge machinery is completely defeated).

I'm happy to raise an issue on the scipy feedstock if it's the right place to initiate the conversation on "how do we do this", but if it seems like this is already on your radar and I'm not sure there's much I can contribute to the solution.

**Comment by h-vetinari** [comment]
If you're talking about MDA in conda-forge, then please don't change the numpy setup on your own. I get that people are eager to change, but none of this is _necessary_ right now. In any case, let's take this discussion elsewhere. Please open an issue on the MDA feedstock (or the numpy one).

**Comment by rgommers** [comment]
I'd leave a change to building against numpy 1.25.x for the next minor release, not a bug fix release. Things are set up well now, and I don't think we have much to gain from doing it in 1.11.2

**Comment by tylerjereddy** [comment]
All 39 CI jobs, including `3.12` wheel builds, are now green, so I'm inclined to merge here pretty soon. I'll probably do another quick scan of the diff, but that's about it unless other concerns are raised soon.

Then I'll start the release process either today or tomorrow.

**Pull request merged**
